docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,242 @@
+[FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ). As well as the
+documentation about [task scheduling (TODO)](<TODO>).
+
+### Task Scheduling And Network Shuffle
+
+Flink jobs consist of different operations that are connected together in a
+dataflow graph. The system decides how to schedule the execution of these
+operations on different processes/machines (TaskManager) and how data is
+shuffled (sent) betwixt them.
+
+Multiple operations/operators can be chained together using a feature called
+[chaining]({% link dev/stream/operators/index.md
+%}#task-chaining-and-resource-groups). A group of one or multiple (chained)
+operators that Flink considers as a unit of scheduling is called a _task_.
+Often the term _subtask_ is used to refer to the individual instances of tasks
+that are running in parallel on multiple TaskManagers but we will only use the
+term _task_ here.
+
+Task scheduling and the network shuffle work different for BATCH execution mode
+and STREAMING execution mode. Mostly due to the fact that we know our input
+data is bounded in BATCH execution mode, which allows Flink to use more
+efficient data structures and algorithms.
+
+We will use this example to explain the differences in task scheduling and
+network transfer:
+
+```java
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+DataStreamSource<String> source = env.fromElements(...);
+
+source.name("source")
+	.map(...).name("map1")
+	.map(...).name("map2")
+	.rebalance()
+	.map(...).name("map3")
+	.map(...).name("map4")
+	.keyBy((value) -> value)
+	.map(...).name("map5")
+	.map(...).name("map6")
+	.sinkTo(...).name("sink");
+```
+
+Operations that imply a 1-to-1 connection pattern between operations, such as
+`map()`, `flatMap()`, or `filter()` can just forward data straight to the next
+operation, which allows these operations to be chained together. This means
+that Flink would not normally insert a network shuffle between them.
+
+Operation such as `keyBy()` or `rebalance()` on the other hand require data to
+be shuffled between different parallel instances of tasks. This induces a
+network shuffle.
+
+For the above example Flink would group operations together as tasks like this:
+
+- Task1: `source`, `map1`, and `map2`
+- Task2: `map3`, `map4`
+- Task3: `map5`, `map6`, and `sink`
+
+And we have a network shuffle between Tasks 1 and 2, and also Tasks 2 and 3.
+
+#### STREAMING Execution Mode
+
+In STREAMING execution mode, all tasks need to be online/running all the time.
+This allows Flink to immediately process new records through the whole
+pipeline, which we need for low-latency stream processing. This also means that
+the TaskManagers that are allotted to a job need to have enough resources to
+run all the tasks at the same time.
+
+Network shuffles are _pipelined_, meaning that records are immediately sent to
+downstream tasks, with some buffering on the network layer. Again, this enables
+low-latency processing.
```

Comment:
It's good the way it is now. Just me nitpicking.

I'd say it is not (only) about low-latency. It's (also) because of the unbounded characteristic of the processing. There is no point in the processing that we can say e.g. an operator will not see any more records and we can drop its state. We must keep its state on all the time because it might change at any point in time in the future.

Reply:
I agree, I think continuous is more important to highlight here than latency

Reply:
I expanded this paragraph. PTAL.